src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -0,0 +1,178 @@
+        /// </remarks>
+        public async Task<ReversedDiagnosticsConnection> AcceptAsync(CancellationToken token)
+        {
+            VerifyNotDisposed();
+
+            using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, _cancellation.Token);
+
+            ReversedDiagnosticsConnection newConnection = null;
+            do
+            {
+                Stream stream = null;
+                IpcAdvertise advertise = null;
+                try
+                {
+                    stream = await _transport.AcceptAsync(linkedSource.Token);
+                    advertise = IpcAdvertise.Parse(stream);
+                }
+                catch (Exception)
+                {
+                    // The advertise data could be incomplete if the runtime shuts down before completely writing
+                    // the information. Catch the exception and continue waiting for a new connection.
+                }
+
+                if (null != advertise)
+                {
+                    Guid runtimeCookie = advertise.RuntimeInstanceCookie;
+                    int pid = unchecked((int)advertise.ProcessId);
+
+                    // If this runtime instance already exists, update the existing connection with the new endpoint.
+                    // Consumers should hold onto the connection instance and use it for diagnostic communication,
+                    // regardless of the number of times the same runtime instance connects. This requires consumers
+                    // to continuously invoke the AcceptAsync method in order to handle runtime instance reconnects,
+                    // even if the consumer only wants to handle a single connection.
+                    ServerIpcEndpoint endpoint = null;
+                    if (!_endpoints.TryGetValue(runtimeCookie, out endpoint))
+                    {
+                        // Create a new endpoint and connection that are cached an returned from this method.
+                        endpoint = new ServerIpcEndpoint();
+                        newConnection = new ReversedDiagnosticsConnection(this, endpoint, pid, runtimeCookie);
+
+                        _endpoints.TryAdd(runtimeCookie, endpoint);
+                    }
+
+                    endpoint.SetStream(stream);
+                }
+            }
+            while (null == newConnection);
+
+            // Send ResumeRuntime message for runtime instances that connect to the server. This will allow
+            // those instances that are configured to wait pause on start to resume after the diagnostics
+            // connection has been made. Instances that are not configured to pause on startup will ignore
+            // the command and return success.
+            var client = new DiagnosticsClient(newConnection.Endpoint);
+            try
+            {
+                client.ResumeRuntime();
```

Comment:
Sending the ResumeRuntime command will use the new connection, thus the endpoint will have a null stream until the runtime reconnects and this AcceptAsync method is called again to fill in the stream. This would likely block the thread that resumed the AcceptAsync continuation IF it immediately attempts to use the endpoint before calling AcceptAsync again.

I'm going to see if I can refactor the AcceptAsync method a bit to make sure that the endpoint is immediately viable when it is returned from AcceptAsync.

Reply:
@josalem, I'm thinking this through a bit more and wanted to hear from you what the intended use of the pause on start behavior is supposed to be on the client side (the process that is not the runtime instance).

As I have it here (when I fix up the issue that I pointed out), the runtime instance associated with the connection will have been already resumed before returning from the AcceptAsync method. Is that intended, or should the client set up (for example) a event pipe session before the runtime instance is resumed so that whatever data is produced by the runtime upfront is collected. I'm not even sure if starting an event pipe session (or any other diagnostic collection) is allowed before the runtime is resumed.

Reply:
What I'm wondering is if we need to let the caller of `ReversedDiagnosticsServer.AcceptAsync` to send the ResumeRuntime command when it is ready to allow the runtime instance to resume rather than the AcceptAsync method automatically doing it. If that's the case, we'd need to document that fairly well and update the `ReversedServerConnectionsSource.ListenAsync` method to send it for each returned connection.

Reply:
The `PauseOnStart` functionality and the corresponding `ResumeRuntime` command are meant to allow users to create EventPipe sessions before the runtime has finished starting.  The stopping point is in runtime bringup before managed code has begun executing and before several major components, e.g., GC, ThreadPool, etc., have started.  The `PauseOnStart` is automatically on if the `DOTNET_DiagnosticsMonitorAddress` config knob is set.  You can set `DOTNET_DiagnosticsMonitorPauseOnStart` to `0` to opt out of that behavior.  I would expect the base API to simply return the stream and allow users to send whatever commands they want before sending the `ResumeRuntime` command.  In the case of `dotnet-monitor` I think it is fine for it just return the stream after sending the resume runtime command.  It should only be sent on the first connection of a runtime instance since you already know it should be in a resumed state after that.